ext/tls/tls_key.rs

```diff
@@ -0,0 +1,308 @@
+use deno_core::futures::future::Either;
+use deno_core::futures::FutureExt;
+use deno_core::unsync::spawn;
+use rustls::ServerConfig;
+use rustls_tokio_stream::ServerConfigProvider;
+use std::cell::RefCell;
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::future::ready;
+use std::future::Future;
+use std::io::ErrorKind;
+use std::rc::Rc;
+use std::sync::Arc;
+use tokio::sync::broadcast;
+use tokio::sync::mpsc;
+use tokio::sync::oneshot;
+
+type ErrorType = Rc<AnyError>;
+
+/// A TLS certificate/private key pair.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct TlsKey(pub Vec<Certificate>, pub PrivateKey);
+
+#[derive(Clone, Debug, Default)]
+pub enum TlsKeys {
+  // TODO(mmastrac): We need Option<&T> for cppgc -- this is a workaround
+  #[default]
+  Null,
+  Static(TlsKey),
+  Resolver(TlsKeyResolver),
+}
+
+pub struct TlsKeysHolder(RefCell<TlsKeys>);
+
+impl TlsKeysHolder {
+  pub fn take(&self) -> TlsKeys {
+    std::mem::take(&mut *self.0.borrow_mut())
+  }
+}
+
+impl From<TlsKeys> for TlsKeysHolder {
+  fn from(value: TlsKeys) -> Self {
+    TlsKeysHolder(RefCell::new(value))
+  }
+}
+
+impl TryInto<Option<TlsKey>> for TlsKeys {
+  type Error = Self;
+  fn try_into(self) -> Result<Option<TlsKey>, Self::Error> {
+    match self {
+      Self::Null => Ok(None),
+      Self::Static(key) => Ok(Some(key)),
+      Self::Resolver(_) => Err(self),
+    }
+  }
+}
+
+impl From<Option<TlsKey>> for TlsKeys {
+  fn from(value: Option<TlsKey>) -> Self {
+    match value {
+      None => TlsKeys::Null,
+      Some(key) => TlsKeys::Static(key),
+    }
+  }
+}
+
+enum TlsKeyState {
+  Resolving(broadcast::Receiver<Result<TlsKey, ErrorType>>),
+  Resolved(Result<TlsKey, ErrorType>),
+}
+
+struct TlsKeyResolverInner {
+  resolution_tx: mpsc::UnboundedSender<(
+    String,
+    broadcast::Sender<Result<TlsKey, ErrorType>>,
+  )>,
+  cache: RefCell<HashMap<String, TlsKeyState>>,
+}
+
+#[derive(Clone)]
+pub struct TlsKeyResolver {
+  inner: Rc<TlsKeyResolverInner>,
+}
+
+impl TlsKeyResolver {
+  async fn resolve_internal(
+    &self,
+    sni: String,
+    alpn: Vec<Vec<u8>>,
+  ) -> Result<Arc<ServerConfig>, AnyError> {
+    let key = self.resolve(sni).await?;
+
+    let mut tls_config = ServerConfig::builder()
+      .with_safe_defaults()
+      .with_no_client_auth()
+      .with_single_cert(key.0, key.1)?;
+    tls_config.alpn_protocols = alpn;
+    Ok(tls_config.into())
+  }
+
+  pub fn into_server_config_provider(
+    self,
+    alpn: Vec<Vec<u8>>,
+  ) -> ServerConfigProvider {
+    let (tx, mut rx) = mpsc::unbounded_channel::<(_, oneshot::Sender<_>)>();
+
+    // We don't want to make the resolver multi-threaded, but the `ServerConfigProvider` is
+    // required to be wrapped in an Arc. To fix this, we spawn a task in our current runtime
+    // to respond to the requests.
+    spawn(async move {
+      while let Some((sni, txr)) = rx.recv().await {
+        _ = txr.send(self.resolve_internal(sni, alpn.clone()).await);
+      }
```

Comment:
Should we store this task somehow? Is it okay if it's "dangling"?

Reply:
These are kind of like the floating promise - they get torn down when the other ends of the queues are completed.

Reply:
That makes sense.